volatility3/framework/layers/intel.py

```diff
@@ -102,70 +104,84 @@ def _translate(self, offset: int) -> Tuple[int, int, str]:
-                                                          f"Page Fault at entry {hex(entry)} in page entry")
+            return self._handle_page_fault(self.name, offset, position + 1, entry, f"Page Fault at entry {hex(entry)} in page entry")
+
         page = self._mask(entry, self._maxphyaddr - 1, position + 1) | self._mask(offset, position, 0)
 
         return page, 1 << (position + 1), self._base_layer
 
-    def _translate_entry(self, offset: int) -> Tuple[int, int]:
+    def _find_structure_index(self, position: int) -> Tuple[str, int, bool]:
+        if not self._structure_position_table:
+            counter = self._initial_position
+            for name, size, large_page in self._structure:
+                self._structure_position_table[counter] = name, size, large_page
+                counter -= size
+            # print(self._structure_position_table)
+        return self._structure_position_table[position]
+
+    def _handle_page_fault(self, name, offset, invalid_bits, entry, description):
+        """Handles page faults"""
+        raise exceptions.PagedInvalidAddressException(self.name, offset, invalid_bits, entry,
+                                                      "Page Fault at entry " + hex(entry) + " in table " + name)
+
+    def _translate_entry(self, offset: int, position: int, entry: int) -> Tuple[int, int]:
         """Translates a specific offset based on paging tables.
 
         Returns the translated entry value
         """
-        # Setup the entry and how far we are through the offset
-        # Position maintains the number of bits left to process
-        # We or with 0x1 to ensure our page_map_offset is always valid
-        position = self._initial_position
-        entry = self._initial_entry
-
-        if self.minimum_address > offset > self.maximum_address:
+        name, size, large_page = self._find_structure_index(position)
+
+        # Check we're present
+        if not entry & 0x1:
+            return self._handle_page_fault(self.name, offset, position + 1, entry,
+                                           "Page Fault at entry " + hex(entry) + " in table " + name)
+        # Check if we're a large page
+        if large_page and (entry & (1 << 7)):
+            # We're a large page, the rest is finished below
+            # If we want to implement PSE-36, it would need to be done here
+            return entry, position
+        # Figure out how much of the offset we should be using
+        start = position
+        position -= size
+        index = self._mask(offset, start, position + 1) >> (position + 1)
+
+        # Grab the base address of the table we'll be getting the next entry from
+        base_address = self._mask(entry, self._maxphyaddr - 1, size + self._index_shift)
+
+        table = self._get_valid_table(base_address)
+        if table is None:
             raise exceptions.PagedInvalidAddressException(self.name, offset, position + 1, entry,
```

Comment:
It seems that this point can also use the `_handle_page_fault` function.

Reply:
replaced with `self._handle_page_fault`